### ISSUE TITLE
I'd like to make this happen

### ISSUE THREAD

**Issue by g-andrade** [sqids/sqids-elixir#1]
Hi,

I'd like to contribute an Elixir implementation of `sqids`.

@4kimov

**Comment by 4kimov** [comment]
Hi @g-andrade, thanks for pinging. Sent you an invite. Spec is [here](https://github.com/sqids/sqids-spec#-porting-to-a-new-language), if you have questions, I'm here 👍

**Comment by g-andrade** [comment]
I got encoding matching [the playground](https://sqids.org/playground) for a few single numbers with default alphabet and blocklist (it fails with more than one number).

![Screenshot from 2023-09-10 20-23-36](https://github.com/sqids/sqids-elixir/assets/7349502/2d6afb92-37cd-4940-a7af-33a1ec7f59ec)

Encouraging! I hope to have a chance to pick it up again soon and start pushing into a draft PR.

**Comment by 4kimov** [comment]
@g-andrade Good job! Btw, while this repo is empty and the code is pre-production (most Sqids versions are v0.x for now), feel free to just push to `main`, no need for a PR

**Comment by 4kimov** [comment]
@g-andrade How's it going? Need help with anything?

**Comment by g-andrade** [comment]
Hi Ivan, I might continue next weekend, if not on another after that. In
the mean time, I’ll push my fork’s feature branch into main.

**Comment by g-andrade** [comment]
I had a lucky break and got to implement most of the spec except for blocklist handling during encoding.

Test cases copied verbatim from spec and passing (except for blocklist):
https://github.com/sqids/sqids-elixir/blob/c2e64ae/test/sqids_test.exs

Things missing:
- [ ] Apply blocklist properly during encoding
- [ ] CI through GitHub Actions on every git push
- [ ] Embedded text descriptions for generated API reference

A few questions:
- When characters are found within an ID that are not part of the alphabet, the reference implementation returns an empty array. Would it be OK for this implementation to fail instead?
- Since Elixir's integers are bigints, they're only constraint by memory (and common sense). Should this implementation limit accepted numbers, both when encoding and/or decoding, to JS's [MAX_SAFE_INTEGER](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_SAFE_INTEGER)? Ideally this implementation wouldn't produce ids others cannot decode. Perhaps limiting can be done by default but disabled through an encoding option.

**Comment by 4kimov** [comment]
Regarding your questions:

1. Up to you, but the reason `decode()` in the spec does not throw an error but returns an empty array, is for consistency. The decoding [does not guarantee to return "correct" numbers](https://sqids.org/faq#valid-ids), the user has to re-encode them manually to check that the ID is "valid" or canonical. If you decide to throw an error in that spot, you might accidentally signal to the user that by returning numbers, they're guaranteed to be valid (which is not the case).

2. Every implementation would ideally return the biggest integer that particular language supports. So for Elixir, you can just support bigints (if that's the biggest) without worrying if other languages can reach those limits or not - that's up to the end user.

**Comment by g-andrade** [comment]
A couple of updates:
* blocklist implemented and test cases added for it
* CI setup on GitHub Actions (and passing) for recent Elixir versions

**Comment by g-andrade** [comment]
Nearly there:
* there's a README (which I based on the one for the Python library)
* there's a function reference
* there's test coverage of all relevant edge cases I could think of

At this point only a final review and the upload of a package to [Hex](https://hex.pm/) are likely missing.

**Comment by 4kimov** [comment]
@g-andrade Cool.

> there's test coverage of all relevant edge cases I could think of

I can see there are ported over tests in this file https://github.com/sqids/sqids-elixir/blob/main/test/sqids_test.exs but not sure if it includes all the [tests from the spec](https://github.com/sqids/sqids-spec/tree/main/tests) (we should include every one because they cover some good edge-cases). Otherwise, any extra tests should be purely Elixir-specific.

Anyway, ready to update the site whenever you are ready with the lib 👍 Let me know if there's anything specific you'd like me to look at

**Comment by g-andrade** [comment]
> […] but not sure if it includes all the [tests from the spec](https://github.com/sqids/sqids-spec/tree/main/tests) (we should include every one because they cover some good edge-cases).

No worries there, I translated the full 4 suites from the spec a few weeks ago. The test cases added since then are mostly Elixir specific, for ex. which exceptions or errors are thrown or returned, when and how - to ensure good UX for consumers of this library. It doesn’t hurt that this process helped uncover some smaller bugs or weird semantics as well, which were fixed or tweaked as a result.

If I find the time next weekend, I’ll polish any details missing and figure out package publishing, and once that’s done we’re good to release 🎉

Going back to nice-to-haves, automated update of the blocklist is coded through a GitHub Actions workflow: daily checks with a PR opened whenever there are changes. It requires setting up some extra permissions in repo settings; I’ll create a ticket for it but it’s probably not a big priority since the blocklist won’t change that often (I assume).

Another optional quality of life improvement would be to enable Dependabot in the repo settings; its config is ready and pushed. Not critical either, since all of the current dependencies are dev-only: for static analysis, debugging or blocklist updates.
